### PR TITLE
feat(sdk): cip 0001 migration v0

### PR DESCRIPTION
**Pull request by rtb-12**
# [sdk] Add CIP-0001 migration v0 support for atomic application state upgrades

## Description

Implements CIP-0001 Phase 7 (Upgrade & Migration) using the embedded migration model (v0). This enables developers to safely upgrade application state schemas during context application updates.

**Key features:**
- **`#[app::migrate]` proc-macro** — Marks migration functions in the new WASM module that read old state via `read_raw()` and return new state
- **AppKey continuity enforcement** — Validates that `signerId` matches between old and new applications before migration, preventing unauthorized hijacking
- **Atomic migration execution** — New state is committed only on success; failures preserve original state
- **CLI support** — New `--migrate-method` flag for `meroctl context update`

**Changes by layer:**
- `crates/sdk/` — Added `read_raw()` state function and `#[app::migrate]` macro
- `crates/prelude/` — Added `ROOT_STORAGE_ENTRY_ID` constant
- `crates/storage/` — Added `root_storage_key()` helper
- `crates/context/` — Added migration execution flow with AppKey continuity checks (16 unit tests)
- `crates/server/` — Plumbed `migrate_method` through API layer
- `crates/meroctl/` — Added `--migrate-method` CLI flag
- `apps/kv-store-v2/` — Example migration app for testing

Closes: Implements CIP-0001 Phase 7 (Migration v0)

## Test plan

- [x] `cargo test -p calimero-storage -- constants` — Root storage key tests
- [x] `cargo test -p calimero-sdk-macros` — Macro expansion tests  
- [x] `cargo test -p calimero-context -- update_application` — 16 unit tests covering:
  - AppKey continuity (matching signerIds pass)
  - Security edge cases (case sensitivity, homoglyph attacks, whitespace)
  - Rollback guarantees (no state written on failure)
- [x] `cargo check --workspace` passes
- [x] `cargo clippy -- -A warnings` passes
- [x] Manual E2E testing available via `apps/kv-store-v2/MANUAL_TESTING.md`:
  ```bash
  meroctl --node <node> app install --path apps/kv-store/res/kv-store-1.0.0.mpk --dev
  meroctl --node <node> context create --application-id <v1_app_id>
  meroctl --node <node> call --context <ctx> --method set --args '{"key":"test","value":"data"}'
  meroctl --node <node> app install --path apps/kv-store-v2/res/kv-store-2.0.0.mpk --dev
  meroctl --node <node> context update --context <ctx> --application-id <v2_app_id> --migrate-method migrate_v1_to_v2
  meroctl --node <node> call --context <ctx> --method get --args '{"key":"test"}'  # Returns preserved data
  ```

**Note:** Automated merobox E2E test (`migration-test.yml`) blocked pending `update_context_application` step type support.

## Documentation update

- `apps/kv-store-v2/MANUAL_TESTING.md` — Manual testing guide (included)
- `apps/kv-store-v2/README.md` — Example app documentation (TODO)
- SDK docs for `#[app::migrate]` and `read_raw()` — Should be updated within one day of merge
- CIP-0001 spec already covers the design in `specs/CIP-0001.md`

---

> [!NOTE]
> **High Risk**
> Adds a new optional migration execution path to context application updates, including direct root-state rewrites and Merkle/index maintenance, so bugs could corrupt context state or break sync if incorrect. Also tightens update authorization via signerId continuity checks, which may block previously-allowed upgrade flows if metadata is inconsistent.
> 
> **Overview**
> Enables **atomic state schema migrations during `context update`** by plumbing an optional `migrate_method` from `meroctl` and the admin API down into `ContextManager` update handling.
> 
> Introduces SDK-level migration building blocks: a new `#[app::migrate]` proc-macro that exports a WASM migration entrypoint and a `read_raw()` helper to fetch prior root-state bytes (plus shared root storage constants via `calimero-prelude`).
> 
> Implements the migration execution flow in `crates/context`: invalidate cached modules when migrating, run the specified migration function in the *new* WASM, write the returned state back to root storage while updating indexes/Merkle root (and clearing pending delta actions), update `root_hash`/`dag_heads`, emit events/logs, and enforce **AppKey continuity** (signerId match; reject signed→unsigned downgrades) with added unit tests.
> 
> Adds a migration example suite (`apps/migrations/migration-suite-v1..v5`), plus build scripts, an E2E migration test script, and new lifecycle documentation.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit ba3c33d463945145787e8b3f48d93e903fe51149. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: crates/prelude/src/constants.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+pub use calimero_primitives::common::DIGEST_SIZE;
+
+/// Well-known ID for the root storage entry.
+///
+/// This is a fixed 32-byte value used as the entry ID for the application's root state.
+/// The value `118` represents the ASCII code for 'v' (value), chosen as a memorable constant.
+///
+/// The actual storage key is computed by hashing this ID with the `Key::Entry` discriminant
+/// prefix, resulting in a unique key for the root state entry.
+pub const ROOT_STORAGE_ENTRY_ID: [u8; DIGEST_SIZE] = [118u8; DIGEST_SIZE];
```
[the remark this answers is not in the document]
Comment (rtb-12) [reply crates/prelude/src/constants.rs:17]: The issue as described is **not valid**. Here's why:

The storage layer has a two-level hierarchy for root state:

```
Collection(Id::root() = context_id())      ← Merkle tree root node
  └── Entry<T>(Id::new([118; 32]))          ← Root state data
```

This is confirmed by `Root::entry_id()` at `crates/storage/src/collections/root.rs:69-71`:

```69:71:crates/storage/src/collections/root.rs
    fn entry_id() -> Id {
        Id::new([118; 32])
    }
```

And `Root::new_internal` uses this exact ID when inserting the entry:

```55:60:crates/storage/src/collections/root.rs
    pub fn new_internal<F: FnOnce() -> T>(f: F) -> Self {
        let mut inner = Collection::new(Some(*ROOT_ID));
        let id = Self::entry_id();
        let value = inner.insert(Some(id), f()).unwrap();
```

So `ROOT_STORAGE_ENTRY_ID = [118u8; 32]` is correct — it matches the hardcoded `entry_id()` in the storage layer.

Regarding the `CannotCreateOrphan` concern, the check in `save_raw` at line 1050 is:

```1050:1052:crates/storage/src/interface.rs
        if !id.is_root() && <Index<S>>::get_parent_id(id)?.is_none() {
            return Err(StorageError::CannotCreateOrphan(id));
        }
```

This only rejects the write if **both** conditions are true: the ID isn't root **AND** it has no parent. During migration, the entry already exists as a child of the root Collection (established during the original `Root::new_internal` call), so `get_parent_id(id)` returns `Some(root_collection_id)` — the orphan check passes.

Similarly, `read_raw()` computes `SHA256(0x01 || [118; 32])` which is exactly the `Key::Entry` for this ID — the same key the normal storage path writes to.

**No fix needed.** The constant is correct and the code works as designed.

**Review by xilosada — APPROVED** [review]
